core/src/main/java/org/elasticsearch/common/util/IndexFolderUpgrader.java

```diff
@@ -0,0 +1,155 @@
+            return;
+        }
+        boolean success = false;
+        try {
+            Files.move(source, target, StandardCopyOption.ATOMIC_MOVE);
+            success = true;
+        } catch (NoSuchFileException ignored) {
+            assert ignored.getFile().equals(source.toAbsolutePath().toString()) : "must be same as the source";
+            // thrown when the source is non-existent because the folder was renamed
+            // by another node (shared FS) after we checked if the target exists
+            assert Files.exists(target) : "must be upgraded by another node";
+        } finally {
+            if (success) {
+                logger.info("{} moved from [{}] to [{}]", index, source, target);
+                logger.trace("{} syncing directory [{}]", index, target);
+                IOUtils.fsync(target, true);
+            }
+        }
+    }
+
+    /**
+     * Renames <code>indexFolderName</code> index folders found in node paths and custom path
+     * iff {@link #needsUpgrade(Index, String)} is true.
+     * Index folder in custom paths are renamed first followed by index folders in each node path.
+     */
+    void upgrade(final String indexFolderName) throws IOException {
+        for (NodeEnvironment.NodePath nodePath : nodeEnv.nodePaths()) {
+            final Path indexFolderPath = nodePath.indicesPath.resolve(indexFolderName);
+            if (Files.exists(indexFolderPath) == false) {
+                continue;
+            }
+            final IndexMetaData indexMetaData = indexStateFormat.loadLatestState(logger, indexFolderPath);
+            if (indexMetaData != null) {
+                final Index index = indexMetaData.getIndex();
+                if (needsUpgrade(index, indexFolderName)) {
+                    logger.info("{} upgrading [{}] to new naming convention", index, indexFolderPath);
+                    final IndexSettings indexSettings = new IndexSettings(indexMetaData, settings);
+                    if (indexSettings.hasCustomDataPath()) {
+                        // we rename index folder in custom path before renaming them in any node path
+                        // to have the index state under a not-yet-upgraded index folder, which we use to
+                        // continue renaming after a incomplete upgrade.
+                        // NOTE: we rename the folder in custom path only the first time we encounter a state
+                        // in a node path, which needs upgrading, it is a no-op for subsequent node paths
+                        upgrade(index, nodeEnv.resolveCustomLocation(indexSettings).resolve(indexFolderName));
+                    }
+                    upgrade(index, indexFolderPath);
+                } else {
+                    logger.info("[{}] no upgrade needed - already upgraded", indexFolderPath);
+                }
+            } else {
+                logger.info("[{}] no index state found - ignoring", indexFolderPath);
```

Comment:
Now we just log it as a warn rather than throw a hard exception in this case, as we might have a node crash as it is [writing the index state](https://github.com/elastic/elasticsearch/blob/master/core/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java#L161). This may leave folders in indices folder with no valid index state. Currently, if a node crashed in such a state before being upgraded, the empty folder is left in the file system (we don't rename). If a node crashed in such a state after being upgraded, the upgrader just ignores it and let cluster state listener deal with it.